sdk/apis/apis/v1alpha2/types_apiexport_conversion.go

```diff
@@ -0,0 +1,186 @@
+		return err
+	}
+	if err := Convert_v1alpha2_APIExportStatus_To_v1alpha1_APIExportStatus(&in.Status, &out.Status, s); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func convert_v1alpha2_ResourceSchemas_To_v1alpha1_LatestResourceSchemas(in APIExportSpec) ([]string, []ResourceSchema) {
+	hubSchemas := []string{}
+	nonCRDSchemas := []ResourceSchema{}
+
+	if schemas := in.ResourceSchemas; schemas != nil {
+		for _, schema := range schemas {
+			if schema.Storage.CRD != nil {
+				hubSchemas = append(hubSchemas, schema.Schema)
+			} else {
+				nonCRDSchemas = append(nonCRDSchemas, schema)
+			}
+		}
+	}
+
+	return hubSchemas, nonCRDSchemas
+}
+
+// Convert_v1alpha2_APIExportSpec_To_v1alpha1_APIExportSpec is *not* lossless, as it will drop all non-CRD
+// resource schemas present in the APIExport's spec. To have a full, lossless conversion, use
+// Convert_v1alpha2_APIExport_To_v1alpha1_APIExport instead.
+func Convert_v1alpha2_APIExportSpec_To_v1alpha1_APIExportSpec(in *APIExportSpec, out *apisv1alpha1.APIExportSpec, s kubeconversion.Scope) error {
+	if in.Identity != nil {
+		out.Identity = &apisv1alpha1.Identity{}
+		if err := Convert_v1alpha2_Identity_To_v1alpha1_Identity(in.Identity, out.Identity, s); err != nil {
+			return err
+		}
+	}
+
+	if in.MaximalPermissionPolicy != nil {
+		out.MaximalPermissionPolicy = &apisv1alpha1.MaximalPermissionPolicy{}
+		if err := Convert_v1alpha2_MaximalPermissionPolicy_To_v1alpha1_MaximalPermissionPolicy(in.MaximalPermissionPolicy, out.MaximalPermissionPolicy, s); err != nil {
+			return err
+		}
+	}
+
+	if claims := in.PermissionClaims; claims != nil {
+		newClaims := []apisv1alpha1.PermissionClaim{}
+		for _, claim := range claims {
+			var newClaim apisv1alpha1.PermissionClaim
+			if err := Convert_v1alpha2_PermissionClaim_To_v1alpha1_PermissionClaim(&claim, &newClaim, s); err != nil {
+				return err
+			}
+			newClaims = append(newClaims, newClaim)
+		}
+		out.PermissionClaims = newClaims
+	}
+
+	latest, _ := convert_v1alpha2_ResourceSchemas_To_v1alpha1_LatestResourceSchemas(*in)
+	if len(latest) > 0 {
+		out.LatestResourceSchemas = latest
+	}
+
+	return nil
+}
+
+func Convert_v1alpha1_APIExport_To_v1alpha2_APIExport(in *apisv1alpha1.APIExport, out *APIExport, s kubeconversion.Scope) error {
+	if err := autoConvert_v1alpha1_APIExport_To_v1alpha2_APIExport(in, out, s); err != nil {
+		return err
+	}
+
+	if overhanging, ok := in.Annotations[resourceSchemasAnnotation]; ok {
+		resourceSchemas := []ResourceSchema{}
+		if err := json.Unmarshal([]byte(overhanging), &resourceSchemas); err != nil {
+			return fmt.Errorf("failed to decode schemas from JSON: %w", err)
+		}
+
+		if len(resourceSchemas) > 0 {
+			if out.Spec.ResourceSchemas == nil {
+				out.Spec.ResourceSchemas = []ResourceSchema{}
+			}
+
+			out.Spec.ResourceSchemas = append(out.Spec.ResourceSchemas, resourceSchemas...)
+		}
+
+		delete(out.Annotations, resourceSchemasAnnotation)
```

Comment:
All these mutations look suspicious. Wondering whether we might cache mutations here.

Look at the invariant defined here:

```
	// Convert attempts to convert one object into another, or returns an error. This
	// method does not mutate the in object, but the in and out object might share data structures,
	// i.e. the out object cannot be mutated without mutating the in object as well.
	// The context argument will be passed to all nested conversions.
	Convert(in, out, context interface{}) error
```

In other words, in and out share data structure. But when you mutate out, you must do a (hopefully shallow, avoid deep copies) copy.

Reply:
This is fuzzing test where we check that we didn't modified the input itself, right? Or you had something else inmind with `cache` ?

Reply:
Right. If you roundtrip test checks the input, it is fine.